include/util/ThreadPool.h

```diff
@@ -0,0 +1,148 @@
+/*******************************************************************************
+* Copyright 2009-2015 Juan Francisco Crespo Galán
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+******************************************************************************/
+
+#pragma once
+
+/**
+* @file ThreadPool.h
+* @ingroup util
+* The ThreadPool class.
+*/
+
+#include "Audaspace.h"
+
+#include <mutex>
+#include <condition_variable>
+#include <vector>
+#include <thread>
+#include <queue>
+#include <future>
+
+AUD_NAMESPACE_BEGIN
+/**
+* This represents pool of threads.
+*/
+class AUD_API ThreadPool
+{
+private:
+	/** 
+	* A queue of tasks.
+	*/
+	std::queue<std::function<void()>> m_queue;
+
+	/**
+	* A vector of thread objects.
+	*/
+	std::vector<std::thread> m_threads;
+
+	/**
+	* A mutex for synchronization.
+	*/
+	std::mutex m_mutex;
+
+	/**
+	* A condition variable used to stop the threads when there are no tasks.
+	*/
+	std::condition_variable m_condition;
+
+	/**
+	* Stop flag.
+	*/
+	bool m_stopFlag;
+
+	/**
+	* The number fo threads.
+	*/
+	unsigned int m_numThreads;
+
+	// delete copy constructor and operator=
+	ThreadPool(const ThreadPool&) = delete;
+	ThreadPool& operator=(const ThreadPool&) = delete;
+public:
+	/**
+	* Creates a new ThreadPool object.
+	* \param count The number of threads of the pool. It must not be 0.
+	*/
+	ThreadPool(unsigned int count) :
+		m_stopFlag(false), m_numThreads(count)
+	{
+		for(unsigned int i = 0; i < count; i++)
```

Comment:
There are many reasons to keep as much code as possible outside of headers:

- compilation time (every cpp including the header needs to compile all of it)
- binary size (the functions are compiled into machine code once for every cpp file and if the linker is not able to optimize that, they are all included in the binary; or it "optimizes" by inlining guaranteeing the unwanted duplication)
- unnecessary recompilation (every time you change something inside the functions that would be in the cpp, all cpps depending on the header get recompiled)
- users of the library not having to worry about the implementation (they just use headers, so they just worry about what is in there)
- the principle of separating interface and implementation (which unfortunately is not easily possible with templates)
- probably there are more reasons